File: toy_measurement.py
import os
import tomllib

import h5py
import numpy as np
import sys
import time
sys.path.append("./BathymetryReco/src")
from swe_wrapper import SWESolver, rampFunc, gaussian_bathymetry


def main():
    """This script is used to generate the simulation data to use
    for reconstruction of bathymetry."""
    # Change to the directory of this file
    os.chdir(os.path.dirname(os.path.abspath(__file__)))
    with open("config.toml" , "rb") as f:
        config = tomllib.load(f)
    path = config["observation"]["path"]

    # Ensure the directory exists
    os.makedirs(path, exist_ok=True)

    # Set up the parameters for measurement simulation (# for simulation in MCMC)
    sim_params = config["simulation"]
    xbounds = sim_params["xbounds"]
    nx =  sim_params["nx"]
    total_t = sim_params["tinterval"]
    tstart = sim_params["tstart"]
    timestep = sim_params["timestep"]
    g = sim_params["g"]
    kappa = sim_params["kappa"]
    dealias = sim_params["dealias"]
    problemtype = sim_params["scenario"]
    problem_bathy = sim_params["bathymetry"]
    left_bc = sim_params["bc_file"]
    store = False

    # Create SWE solver and calculate the solution
    solver = SWESolver(xbounds, timestep, nx, total_t,
                        tstart=tstart, g=g, kappa=kappa, dealias=dealias,
                        problemtype=problemtype, bc_file=left_bc)
    x = solver.domain.x

    # Set up the bathymetry
    bathy_config = config["bathymetry"]
    bathy_params = bathy_config["parameters"]

    if bathy_config["parametrized"]:
        assert bathy_config["npeaks"] == len(bathy_params)//2 # Check if the number of peaks is correct
        bathy = np.zeros_like(x)
        for i in range(bathy_config["npeaks"]):
            bathy = gaussian_bathymetry(x, bathy_params[2*i:2*(i+1)])
    else:
        bathy = rampFunc(x)

    print("Start Simulation")
    start = time.time()
    H_sensor, t_array, h_array, u_array = solver.solve(bathy)
    end = time.time()
    print(f"Simulation Done after {end-start}")
    dx = (xbounds[1]-xbounds[0])/nx
    # filename
    filename = f"simulation_data_{problemtype}_{problem_bathy}_test.h5"
    full_path = os.path.join(path, filename)
    print("Saving results to: ", full_path)
    if store:
        with h5py.File(full_path, "w") as f:
            f.create_dataset("h", data=h_array)
            f.create_dataset("u", data=u_array)
            f.create_dataset("H_sensor", data=H_sensor)
            f.create_dataset("b_exact", data=bathy)
            f.create_dataset("xgrid", data=np.copy(solver.domain.x))
            f.create_dataset("t_array", data=t_array)
            f.attrs["T_N"] = total_t
            f.attrs["xmin"] = xbounds[0]
            f.attrs["xmax"] = xbounds[1]
            f.attrs["tstart"] = tstart
            f.attrs["dt"] = timestep
            f.attrs["dx"] = dx
            f.attrs["g"] = g
            f.attrs["k"] = kappa
            f.attrs["M"] = nx
            if bathy_config["parametrized"]:
                f.attrs["npeaks"] = bathy_config["npeaks"]
                f.attrs["bathy_params"] = bathy_params

if __name__ == "__main__":
    main()

File: toy_measurement.jl
using Pkg
Pkg.activate(".")
using TOML
using HDF5
using BathymetryReco
using Distributions
using Plots
using DataInterpolations

cd(@__DIR__)

#TODO pass config file as command line argument
config_file = ARGS[1]
config = TOML.parsefile(config_file)
sim_config = read_simulation_parameters(config["simulation"])

xbounds = sim_config.xbounds
nx =  sim_config.nx
total_t = sim_config.tinterval
tstart = sim_config.tstart
timestep = sim_config.timestep
g = sim_config.g
kappa = sim_config.kappa
dealias = sim_config.dealias
problemtype = sim_config.scenario
problem_bathy = sim_config.bathy_name
bc_file = sim_config.bc_file

solver = swe_solver(sim_config);

x = solver.domain.x

# Set up the bathymetry
bathy_config = read_bathymetry_parameters(config["bathymetry"])
bathy_params = bathy_config.θ
npeaks = bathy_config.n_peaks

bathy = zero(x)

if problem_bathy == "gaussian"
    param_tuples =  length(bathy_params)/2 == npeaks ?
    [[bathy_params[i], bathy_params[i+1]] for i in 1:2:length(bathy_params)] : println("Number of peaks and number of parameters doesn't match!")
    for params in param_tuples
       global bathy += bathymetry(x, params)
    end
elseif  problem_bathy == "exact_bathy"
    global bathy = exp_bathymetry(x)
elseif problem_bathy == "random"
    kernel = SqExpMvNormal(nx, 4, 0.005)
    bathy_dist = MvNormal(kernel)
    equi_x = collect(range(xbounds[1], xbounds[2], nx))
    sample_bathy = rand(bathy_dist)
    global pb = plot(equi_x, sample_bathy, label="sample bathymetry")
    global bathy = PCHIPInterpolation(sample_bathy, equi_x)(x)
else
    println("Requested bathymetry type not available")
end

println("Start Simulation")
H_sensor, t_array, h_array, u_array = solver.solve(bathy, sensor_pos=sim_config.sensor_pos)
println("Simulation Done")
dx = (xbounds[2]-xbounds[1])/nx


io_config = read_io_settings(config["output"])
path = io_config.output_dir

if io_config.save
    mkpath(path)
    sim_name = "$(problemtype)_$(problem_bathy)"
    if problem_bathy == "gaussian"
        sim_name *= "_$(npeaks)_peaks"
    end

    sim_path = joinpath(path,sim_name)
    mkpath(sim_path)
    cd(sim_path)

    if problem_bathy == "gaussian"
        sim_details_path = "$bathy_params"
        sim_details_path = "$bathy_params"
        mkpath(sim_details_path)
        cd(sim_details_path)
        savefig(pb, "sample_bathymetry.png")
        savefig(pb, "sample_bathymetry.pdf")
    end

    println("Storing results in: $(pwd())")

    h5open("jl_simulation_data.h5", "w") do file
        file["h"] = h_array
        file["u"] = u_array
        file["H_sensor"] = H_sensor
        file["b_exact"] = bathy
        file["xgrid"] = x
        file["t_array"] = t_array
        attributes(file)["T_N"] = total_t
        attributes(file)["xmin"] = xbounds[1]
        attributes(file)["xmax"] = xbounds[2]
        attributes(file)["tstart"] = tstart
        attributes(file)["dt"] = timestep
        attributes(file)["dx"] = dx
        attributes(file)["g"] = g
        attributes(file)["kappa"] = kappa
        attributes(file)["M"] = nx
        if problem_bathy == "gaussian"
            attributes(file)["npeaks"] = npeaks
            attributes(file)["bathy_params"] = bathy_params
        end
    end
    open("simulation_config.toml", "w") do io
        TOML.print(io, config)
    end
end
